### DISCUSSION TITLE
How do you achieve this?

### DISCUSSION THREAD

**Discussion by Time-Coder** [gaogaotiantian/watchpoints#32] [General]
That's a wonderful tool!! But that's just a magic for me to understand how you monitor the changable of variable(source code is a little complecated for me). 

Do you continuously reading varaible value in a thread or someting else?

**Comment by gaogaotiantian** [comment]
watchpoints is filled with Python black magic, it was not a very good place to start if you are a beginner. However, there are definitely things that are easy to understand. Starthing with ``sys.settrace``, which allows you to trigger a user-defined callback after executing every line(or bytecode). This function is commonly used for debugging tools and some profiling tools (coveragepy for example). I would assume this solves the first question you would have, which is how I could monitor on values. However, the monitoring itself also has some black magic behind the scene, a simple callback after executing a line is probably good enough within the frame(still, with some black magic), but scope crossing brings other issues.

I agree the source of watchpoints is not the easiest to read, but it's partially due to the nature of the project - it's not an application or a framework, it's a tool to solve mystery :)

**Reply by anzhi0708** [reply]
Your Bilibili followers need a tutorial episode about the `ast` module ;-P

**Comment by rsemenoff** [comment]
I have an app that i converted to flask, and recently added flask-cache to avoid data disappearing between ajax calls, but I dont entirely understand the flask-cache documentation. Pandas is also giving me spurious 'set with copy' error/warning. I also know Pandas made undocumented changes to certain functions I had been using... So I'm dependent on libraries with potentially untrustworthy memory management.
    On Friday, March 18, 2022, 12:43:40 AM PDT, Time-Coder ***@***.***> wrote:  
 
 


That's a wonderful tool!! But that's just a magic for me to understand how you monitor the changable of variable(source code is a little complecated for me).

Do you continuously reading varaible value in a thread or someting else?

—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by rsemenoff** [comment]
Actually I kind of assumed this is entirely black magic :) I don't know how it works otherwise...but my real problem is my own garbage code that I was trying to avoid refactoring. In the end I fixed my buggy code the old-old-fashioned way.
    On Friday, March 18, 2022, 11:20:17 AM PDT, gaogaotiantian ***@***.***> wrote:  
 
 


watchpoints is filled with Python black magic, it was not a very good place to start if you are a beginner. However, there are definitely things that are easy to understand. Starthing with sys.settrace, which allows you to trigger a user-defined callback after executing every line(or bytecode). This function is commonly used for debugging tools and some profiling tools (coveragepy for example). I would assume this solves the first question you would have, which is how I could monitor on values. However, the monitoring itself also has some black magic behind the scene, a simple callback after executing a line is probably good enough within the frame(still, with some black magic), but scope crossing brings other issues.

I agree the source of watchpoints is not the easiest to read, but it's partially due to the nature of the project - it's not an application or a framework, it's a tool to solve mystery :)

—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.
